### PR TITLE
Custom CSS In Customizer

### LINKED ISSUE #1621: Custom CSS: add to Customizer / merge with WordPress Core
**Synopsis:** Custom CSS is being added to WordPress Core, and will ship with 4.7, scheduled to be released at the end of the year. Jetpack's Custom CSS module needs to be compatible with that new core Feature.

**Responsible:** TBD

**User experience:** We'll need to figure out a way to:

- Be compatible with the Core CSS editor.
- Stay compatible with existing Jetpack Custom CSS users.
- Stay compatible with themes and plugins using Custom CSS libs (e.g. [Semicolon](https://konstantin.blog/themes/semicolon/) or [Umbra](https://themes.redradar.net/blog/2014/12/08/introducing-umbra/))
- Be compatible with transfers from WordPress.com to a self-hosted WordPress site.

- Related Core ticket: https://core.trac.wordpress.org/ticket/38672
- Main Core CSS ticket: https://core.trac.wordpress.org/ticket/35395
- In Progress PR to add Jetpack's Custom CSS to the customizer. #3152
- Related issues: #1951, #1231

**Complete:** 0%

**Last Updated:** 2016-11-04 (by @jeherve)

**Original issue content:**

> I'd really like to see the custom css module added to the customizer. This works great on wordpress.com so hopefully some of the code can be shared from there. Using it in the Customizer makes using the custom css module fun (since you see instant updates) - whereas using it in a separate page is just a bit of a pain.

### PR DESCRIPTION
**Pull request by georgestephanis**
Will still need file sync configured, and `CustomDesign::is_active_or_preview()` abstracted to a filter.

Fixes #1621

### CODE REVIEW

**Comment by jeherve** [comment]
Existing filters will also need some documentation love.

File: modules/custom-css/custom-css-in-customizer.php [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+	static $setting_order = array();
+
+	public static function init() {
+	//	if ( ! CustomDesign::is_active_or_preview() ) {
+	//		return;
+	//	}
+
+		if ( ! apply_filters( 'enable_custom_customizer', true ) ) {
+			return;
+		}
```
Comment (mattwiebe) [inline modules/custom-css/custom-css-in-customizer.php]: Should consolidate this filter and `CustomDesign::is_active_or_preview()` as hooks onto a new filter like `jetpack_enable_css_in_customizer`

File: modules/custom-css/custom-css-in-customizer.php [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+	}
+
+	/**
+	 * JS needed for the admin portion of the Customizer.
+	 */
+	public static function customize_controls_enqueue_scripts() {
+		wp_register_style( 'jetpack-css-codemirror', plugins_url( 'custom-css/css/codemirror.css', __FILE__ ), array(), '20120905' );
+		wp_enqueue_style( 'jetpack-css-customizer-control', plugins_url( 'custom-css/css/customizer-control.css', __FILE__ ), array( 'jetpack-css-codemirror' ), '20140728' );
+
+		wp_register_script( 'jetpack-css-codemirror', plugins_url( 'custom-css/js/codemirror.min.js', __FILE__ ), array(), '3.16', true );
```
Comment (mattwiebe) [inline modules/custom-css/custom-css-in-customizer.php]: Let's make sure our CodeMirror version is up-to-date.

File: modules/custom-css/custom-css/js/customizer-control.js [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+			// when the CodeMirror instance changes, mirror to the textarea,
+			// where we have our "true" change event handler bound. This allows both to function.
+			this.editor.on( 'change', _.bind( function( editor ) {
+				this._updating = true;
+				this.$input.val( editor.getValue() ).trigger( 'change' );
+				this._updating = false;
+
+				if ( ! edited ) {
+					window.ga && window.ga( 'send', 'event', 'Customizer', 'Typed Custom CSS' );
```
Comment (mattwiebe) [inline modules/custom-css/custom-css/js/customizer-control.js]: Let's remove any `window.ga` things. We can emit events on `window.wp.customize.trigger( 'someEvent' )` and hook into that in a dotcom scrpit.

Comment (georgestephanis) [inline modules/custom-css/custom-css/js/customizer-control.js]: Any chance I can impose on you for help with that?  It's a bit out of my wheelwell.

File: modules/custom-css/custom-css-in-customizer.php (hunk 2 above)
Comment (georgestephanis) [inline modules/custom-css/custom-css-in-customizer.php]: Current is 5.9.0, looks like the 3.16 release was dated Aug 21, 2013 -- https://github.com/codemirror/CodeMirror/tree/3.16.0

That's a pretty big jump, and one I'm not sure I'd be comfortable doing without digging deeper into this than I have time to.  (atm, I have no idea how codemirror works)

Comment (georgestephanis) [inline modules/custom-css/custom-css-in-customizer.php]: In either case, updating CodeMirror would probably be better done as a separate pull request after shipping this.

File: modules/custom-css/custom-css-in-customizer.php (hunk 1 above)
Comment (georgestephanis) [inline modules/custom-css/custom-css-in-customizer.php]: Yeah, the question is just where in the wpcom architecture that should live. :\

File: modules/custom-css/custom-css-in-customizer.php (hunk 2 above)
Comment (mattwiebe) [inline modules/custom-css/custom-css-in-customizer.php]: Agreed, it can wait.

**Comment by kraftbj** [comment]
@georgestephanis @mattwiebe -- What can we do to move this forward? I'm afraid it is out of my wheelhouse to chip in directly.

**Comment by jeherve** [comment]
#5539

**Comment by samhotchkiss** [comment]
Closing in favor of #5527
